docs/architecture/adr-054-protobuf-semver-compat-codegen.md

```diff
@@ -0,0 +1,248 @@
+libraries.**
+
+#### Pinned proto images
+
+As is described in  https://github.com/cosmos/cosmos-sdk/discussions/10582), we
+would need to pin proto images for state machines in the codec registry
+to ensure proper unknown field filtering. This isn't a deal breaker, but it is a bit complex for people to understand
+and requires both:
+
+* an additional build step to configure which may be hard to understand
+* fairly complex linting of pinned proto images vs runtime generated
+  code: https://github.com/cosmos/cosmos-sdk/blob/fdd3d07a28f662c2fabd6007a4a1f64be3a373b3/proto/cosmos/app/v1alpha1/module.proto#L48-L83
+
+#### Complex refactoring of interfaces
+
+As described in https://github.com/cosmos/cosmos-sdk/discussions/10368, all
+interface methods on generated code would need to be removed and refactored into
+a handler approach. It is a complex refactoring especially of `ValidateBasic`
+and alignment on an ideal replacement pattern hasn't been reached yet
+(https://github.com/cosmos/cosmos-sdk/pull/11340).
+
+#### Potential limitations to generated code
+
+In doing benchmarking of ORM generated code several potential optimizations have emerged and obviously the most
+performant optimization would be to generate as much code as possible. But because ORM table descriptors may evolve from
+one version to the next with new indexes and fields being added, this makes generated code state machine breaking.
+Recall that the API module paradigm requires no state machine breaking logic in generated code.
+
+## Decision
+
+To address these issues, we will adopt a hybrid approach which requires changes
+to the https://github.com/cosmos/cosmos-proto code generator.
+
+For client facing code, the API module pattern is still useful given that
+care is taken to only publish stable tags when packages are actually stable
+and in production.
+
+For state machine code, the recommended solution is to generate protobuf
+code used by the module in an `internal` package and modify the code generator
+to *skip global registration* for generated code in internal packages.
+
+Because there are plenty of cases where we want one module to essentially
+be a client of another module, either in `Keeper` method or as proposed in
+[ADR 033](./adr-033-protobuf-inter-module-comm.md), modules will need
+to use different generated code - for instance from an API module - than
+the code used in the state machine itself.
+
+To address this, we will also modify the code generator to generate interfaces
+which allow different concrete structs to be passed into keeper methods and
+ADR 033 than those used internally by the module.
+
+For instance, for a struct `MsgSend`, getters like `GetAmount`, `GetFromAddress`, are already defined so codegen could
+create an interface:
+
+```go
+type MsgSendI interface {
+  protoreflect.ProtoMessage
+  GetFromAddress() string
+  GetToAddress() string
+  GetAmount() []v1beta1.CoinI
```

Comment:
Because the idea is to move towards interfaces where possible in this codegen. Changing the proto type (say to v1.CoinI) would be breaking which is one thing we're trying to avoid with semver. It _should_ be in a differente semver

Reply:
I'm sorry, I'm not following. Isn't this a type of `"foo"`? So I don't see how you can have a version specific type here.